plan/cbo_test.go

```diff
@@ -64,6 +64,51 @@ func (s *testAnalyzeSuite) TestCBOWithoutAnalyze(c *C) {
 	))
 }
 
+func (s *testAnalyzeSuite) TestStraightJoin(c *C) {
+	defer testleak.AfterTest(c)()
+	store, dom, err := newStoreWithBootstrap()
+	c.Assert(err, IsNil)
+	testKit := testkit.NewTestKit(c, store)
+	defer func() {
+		dom.Close()
+		store.Close()
+	}()
+	testKit.MustExec("use test")
+	h := dom.StatsHandle()
+	for _, tblName := range []string{"t1", "t2", "t3", "t4"} {
+		testKit.MustExec(fmt.Sprintf("create table %s (a int)", tblName))
+		c.Assert(h.HandleDDLEvent(<-h.DDLEventCh()), IsNil)
+	}
+
+	testKit.MustQuery("explain select straight_join * from t1, t2, t3, t4").Check(testkit.Rows(
```

Comment:
The two test case could not trigger any join reordering operation even without the straight-join hint. So could you add a test case like: `select straight_join * from t1, t2, t3, t4 on t1.c = t4.c;`?

Reply:
it will trigger join reorder, in tidb master branch:
```sql
TiDB(localhost) > explain select * from t1, t2, t3, t4;
+-----------------+-----------------+---------------------------------+------+-----------------------------------------------+----------------------+
| id              | parents         | children                        | task | operator info                                 | count                |
+-----------------+-----------------+---------------------------------+------+-----------------------------------------------+----------------------+
| TableScan_17    |                 |                                 | cop  | table:t1, range:[-inf,+inf], keep order:false | 10000.00             |
| TableReader_18  | HashLeftJoin_15 |                                 | root | data:TableScan_17                             | 10000.00             |
| TableScan_19    |                 |                                 | cop  | table:t2, range:[-inf,+inf], keep order:false | 10000.00             |
| TableReader_20  | HashLeftJoin_15 |                                 | root | data:TableScan_19                             | 10000.00             |
| HashLeftJoin_15 | HashLeftJoin_13 | TableReader_18,TableReader_20   | root | inner join, inner:TableReader_20              | 100000000.00         |
| TableScan_23    |                 |                                 | cop  | table:t3, range:[-inf,+inf], keep order:false | 10000.00             |
| TableReader_24  | HashLeftJoin_21 |                                 | root | data:TableScan_23                             | 10000.00             |
| TableScan_25    |                 |                                 | cop  | table:t4, range:[-inf,+inf], keep order:false | 10000.00             |
| TableReader_26  | HashLeftJoin_21 |                                 | root | data:TableScan_25                             | 10000.00             |
| HashLeftJoin_21 | HashLeftJoin_13 | TableReader_24,TableReader_26   | root | inner join, inner:TableReader_26              | 100000000.00         |
| HashLeftJoin_13 |                 | HashLeftJoin_15,HashLeftJoin_21 | root | inner join, inner:HashLeftJoin_21             | 10000000000000000.00 |
+-----------------+-----------------+---------------------------------+------+-----------------------------------------------+----------------------+
11 rows in set (0.00 sec)
```

```
r1 = t1 join t2
r2 = t3 join t4
r = r1 join r2
```

in this pr:
```sql
TiDB(localhost) > explain select straight_join * from t1, t2, t3, t4;
+-----------------+-----------------+--------------------------------+------+-----------------------------------------------+----------------------+
| id              | parents         | children                       | task | operator info                                 | count                |
+-----------------+-----------------+--------------------------------+------+-----------------------------------------------+----------------------+
| TableScan_16    |                 |                                | cop  | table:t1, range:[-inf,+inf], keep order:false | 10000.00             |
| TableReader_17  | HashLeftJoin_14 |                                | root | data:TableScan_16                             | 10000.00             |
| TableScan_18    |                 |                                | cop  | table:t2, range:[-inf,+inf], keep order:false | 10000.00             |
| TableReader_19  | HashLeftJoin_14 |                                | root | data:TableScan_18                             | 10000.00             |
| HashLeftJoin_14 | HashLeftJoin_12 | TableReader_17,TableReader_19  | root | inner join, inner:TableReader_19              | 100000000.00         |
| TableScan_20    |                 |                                | cop  | table:t3, range:[-inf,+inf], keep order:false | 10000.00             |
| TableReader_21  | HashLeftJoin_12 |                                | root | data:TableScan_20                             | 10000.00             |
| HashLeftJoin_12 | HashLeftJoin_10 | HashLeftJoin_14,TableReader_21 | root | inner join, inner:TableReader_21              | 1000000000000.00     |
| TableScan_22    |                 |                                | cop  | table:t4, range:[-inf,+inf], keep order:false | 10000.00             |
| TableReader_23  | HashLeftJoin_10 |                                | root | data:TableScan_22                             | 10000.00             |
| HashLeftJoin_10 |                 | HashLeftJoin_12,TableReader_23 | root | inner join, inner:TableReader_23              | 10000000000000000.00 |
+-----------------+-----------------+--------------------------------+------+-----------------------------------------------+----------------------+
```

```
r1 = t1 join t2
t2 = r1 join t3
r = r2 join t4
```

actually join reorder will only be triggered when all the join is inner join and without any `on` or `using` conditions for master and this pr.

in master:
```sql
TiDB(localhost) > explain select * from t1 join t2 join t3 join t4 on t1.a=t4.a;
+-----------------+-----------------+--------------------------------+------+--------------------------------------------------------------------+------------------+
| id              | parents         | children                       | task | operator info                                                      | count            |
+-----------------+-----------------+--------------------------------+------+--------------------------------------------------------------------+------------------+
| TableScan_18    |                 |                                | cop  | table:t1, range:[-inf,+inf], keep order:false                      | 10000.00         |
| TableReader_19  | HashLeftJoin_16 |                                | root | data:TableScan_18                                                  | 10000.00         |
| TableScan_20    |                 |                                | cop  | table:t2, range:[-inf,+inf], keep order:false                      | 10000.00         |
| TableReader_21  | HashLeftJoin_16 |                                | root | data:TableScan_20                                                  | 10000.00         |
| HashLeftJoin_16 | HashLeftJoin_14 | TableReader_19,TableReader_21  | root | inner join, inner:TableReader_21                                   | 100000000.00     |
| TableScan_22    |                 |                                | cop  | table:t3, range:[-inf,+inf], keep order:false                      | 10000.00         |
| TableReader_23  | HashLeftJoin_14 |                                | root | data:TableScan_22                                                  | 10000.00         |
| HashLeftJoin_14 | HashLeftJoin_12 | HashLeftJoin_16,TableReader_23 | root | inner join, inner:TableReader_23                                   | 1000000000000.00 |
| TableScan_24    |                 |                                | cop  | table:t4, range:[-inf,+inf], keep order:false                      | 10000.00         |
| TableReader_25  | HashLeftJoin_12 |                                | root | data:TableScan_24                                                  | 10000.00         |
| HashLeftJoin_12 |                 | HashLeftJoin_14,TableReader_25 | root | inner join, inner:TableReader_25, equal:[eq(test.t1.a, test.t4.a)] | 1250000000000.00 |
+-----------------+-----------------+--------------------------------+------+--------------------------------------------------------------------+------------------+
11 rows in set (0.00 sec)
```